vllm/model_executor/layers/quantization/mxfp4.py

```diff
@@ -33,33 +34,73 @@
-            "For faster performance, consider setting "
-            "VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8=1, "
-            "though this may impact accuracy.")
-        return True
-
-    return False
-
-
-def _should_use_flashinfer_mxfp4_mxfp8():
-    """Determine if FlashInfer MXFP4 MXFP8 should be used."""
-    return envs.VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8
+# enum for mxfp4 backend
+class Mxfp4Backend(Enum):
+    NONE = 0
+
+    # FlashInfer Backend
+    SM100_FI_MXFP4_MXFP8_TRTLLM = 1
+    SM100_FI_MXFP4_MXFP8_CUTLASS = 2
+    SM100_FI_MXFP4_BF16 = 3
+    SM90_FI_MXFP4_BF16 = 4
+
+    # Marlin Backend
+    MARLIN = 5
+
+    # Triton Backend
+    TRITON = 6
+
+
+def get_mxfp4_backend():
+    # Backend Selection
+    if current_platform.is_cuda():
+        if (current_platform.is_device_capability(90) and has_flashinfer()
+                and envs.VLLM_USE_FLASHINFER_MOE_MXFP4_BF16):
+            logger.info_once("Using FlashInfer MXFP4 BF16 backend for SM90")
+            return Mxfp4Backend.SM90_FI_MXFP4_BF16
+        elif (current_platform.is_device_capability(100) and has_flashinfer()
+              and envs.VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8_CUTLASS):
+            logger.info_once(
+                "Using FlashInfer MXFP4 MXFP8 CUTLASS backend for SM100")
+            return Mxfp4Backend.SM100_FI_MXFP4_MXFP8_CUTLASS
+        elif (current_platform.is_device_capability(100) and has_flashinfer()
+              and envs.VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8):
+            logger.info_once(
+                "Using FlashInfer MXFP4 MXFP8 TRTLLM backend for SM100, "
+                "for high concurrency throughput workloads consider setting "
+                "VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8_CUTLASS=1 for better "
+                "performance")
+            return Mxfp4Backend.SM100_FI_MXFP4_MXFP8_TRTLLM
+        elif current_platform.is_device_capability(100) and has_flashinfer():
+            logger.info_once(
+                "Using FlashInfer MXFP4 BF16 backend for SM100, "
+                "For faster performance on SM100, consider setting "
+                "VLLM_USE_FLASHINFER_MOE_MXFP4_MXFP8=1, though this may impact "
+                "accuracy.")
+            return Mxfp4Backend.SM100_FI_MXFP4_BF16
+        elif ((current_platform.is_device_capability(100)
+               or current_platform.is_device_capability(90))
+              and not has_flashinfer()):
+            logger.warning_once(
+                "MXFP4 MoE is enabled on Blackwell and Hopper but FlashInfer "
+                "is not available. This may result in degraded performance. "
+                "Please `pip install vllm[flashinfer]` for best results.")
 
+        # If FlashInfer is not available, try either Marlin or Triton
+        if current_platform.get_device_capability()[0] < 9:
+            logger.info_once("Using Marlin backend")
+            return Mxfp4Backend.MARLIN
+        elif has_triton_kernels():
```

Comment:
This doesn't match the original logic and breaks GPT-OSS + Hopper when no env vars are set. The original logic says that on Hopper GPUs, if triton kernels are not available, it always uses Marlin kernels. But here it will fall through to Mxfp4Backend.NONE and trigger the assertion.

Suggested patch: https://github.com/vllm-project/vllm/commit/5757ff093cf02a33b9000ecf5f3a28a015183e96